### DISCUSSION TITLE
HTML forms with fastapi

### DISCUSSION THREAD

**Discussion by aahnik** [fastapi/fastapi#3007] [Questions]
Inside the docs, we get beautiful HTML forms auto-generated.

What I mean is this 👇

![image](https://user-images.githubusercontent.com/66209958/112757439-69da0a00-9007-11eb-9e83-521723bbb3cc.png)

Is there any library for generating these kinds of forms in HTML?

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Sample Form</title>
</head>
<body>

{{something.form}}

<p>Result: {{ result }}</p>

</body>
</html>
```

the `something.form` is the form automatically generated from the datatypes defined. So that I dont need to write the form manually.

example manual form 👇

```html
<form method="post">
    <input type="number" name="num" value="1234"/>
    <input type="submit">
</form>
```

I am actually talking about a feature like [Django Forms](https://docs.djangoproject.com/en/3.1/topics/forms/#the-template)

**Comment by cosmos369** [comment]
good doubt...somebody pls answer

**Comment by Mause** [comment]
Have you looked at [Wtforms](https://wtforms.readthedocs.io/en/2.3.x/)?

**Reply by imiten** [reply]
I have same query and am afraid answer is what @Mause replied.  For non-json web application (old style web apps) which deal with just html templates in response and serve html back seems the fastapi validation feature using pydantic model or just method parameters model raise RequestValidationError with json exception.  For pure HTML based web apps its not suitable and we have to go using Wtforms or not use validation mechanisms and just code error handling like in old jsp style looking for errors passed in context and show them in template.

**Reply by imiten** [reply]
@Mause how do pass Request to wftforms Form object.  I am getting error: TypeError: formdata should be a multidict-type wrapper that supports the 'getlist' method.  Any pointer will also help.

**Reply by imiten** [reply]
was able to make it work.
```
async def ...:
    formdata=await request.form()
    GameForm(formdata=formdata) 
```

**Comment by lesnik512** [comment]
There are solutions on JavaScript. For example [this](https://github.com/jsonform/jsonform)

Looks like you need to use json schema generated from Pydantic's models by `schema()` or `schema_json()` ([here](https://pydantic-docs.helpmanual.io/usage/schema/))

**Reply by ghandic** [reply]
@aahnik was this answer useful? If so feel free to mark it as answered by @lesnik512  😄
